src/transaction/atomic_replication_helper.cpp

```diff
@@ -301,160 +354,465 @@ namespace cublog
   {
-    assert (m_is_sysop);
-    assert (!m_postpone_started);
-
-    m_postpone_started = true;
+    // TODO: start LSA can also be obtained from the first log entry
+    return m_start_lsa;
   }
 
-  bool
-  atomic_replication_helper::atomic_log_sequence::is_postpone_sequence_started () const
+  void
+  atomic_replication_helper::atomic_log_sequence::append_control_log (LOG_RECTYPE rectype, LOG_LSA lsa)
   {
-    //assert ((m_postpone_started && m_is_sysop) || !m_is_sysop);
+    m_log_vec.emplace_back (lsa, rectype);
 
-    return m_postpone_started;
+    if (prm_get_bool_value (PRM_ID_ER_LOG_PTS_ATOMIC_REPL_DEBUG))
+      {
+	dump ("sequence::append_control_log END");
+      }
   }
 
   void
-  atomic_replication_helper::atomic_log_sequence::complete_one_postpone_sequence ()
+  atomic_replication_helper::atomic_log_sequence::append_control_log_sysop_end (
+	  LOG_LSA lsa, LOG_SYSOP_END_TYPE sysop_end_type, LOG_LSA sysop_end_last_parent_lsa)
   {
-    assert (m_is_sysop);
-    assert (m_postpone_started);
+    m_log_vec.emplace_back (lsa, sysop_end_type, sysop_end_last_parent_lsa);
 
-    ++m_end_pospone_count;
+    if (prm_get_bool_value (PRM_ID_ER_LOG_PTS_ATOMIC_REPL_DEBUG))
+      {
+	dump ("sequence::append_control_log_sysop_end END");
+      }
   }
 
   bool
-  atomic_replication_helper::atomic_log_sequence::is_at_least_one_postpone_sequence_completed () const
+  atomic_replication_helper::atomic_log_sequence::all_log_entries_are_control () const
   {
-    if (m_end_pospone_count > 0)
+    return std::all_of (m_log_vec.cbegin (), m_log_vec.cend (),
+			[] (const atomic_log_entry &entry)
+    {
+      return entry.is_control ();
+    });
+  }
+
+  bool
+  atomic_replication_helper::atomic_log_sequence::can_purge ()
+  {
+    assert (all_log_entries_are_control ());
+
+    // - after the actual payload log records in a sequences have been applied
+    //  this function tries to remove - in a consistent and controlled manner(*) -
+    //  the remaining "control" log recods;
+    // - the logic here is akin to "dynamic programming" as with each new processed
+    //  log record, the state is re-evaluated; similarly, with each applied sequence
+    //  of consecutive payload log records, the remaining log records are re-evaluated
+    //  in this function and, if possible, removed (most of the times in pairs - start -end)
+    // - another benefit is that it allows for "nested" atomic replication sequences with
+    //  [almost] arbitrary structure to be processed in a controlled way:
+    //
+    // Example:
+    //
+    //  ------------------------------------ transactional log axis ----->
+    //  C1  W1 W2 W3  C2  W4 W5 W6  C3  W7 W8 W9 W10 C4  W11 W12  C5  C6
+    //   |             |             |                |            |   |
+    //  start        start           |              start          |   |
+    //                              end                          end   end
+    //
```

Comment:
If my understanding is correct, this modification changes the mechanism that was originally proposed.
1. fix all page
2. apply all and unfix all page.

Found issues are resolved by this fix. (I think)
1. vacuum - postpone dealloc
2. update - alloc new heap page for newhome

No problematic cases were found, but the reflection of W1 W2 W3 and W7 W8 W9 W10 is no longer atomic.

Reply:
> but the reflection of W1 W2 W3 and W7 W8 W9 W10 is no longer atomic

Correct. We will continue to investigate and evaluate what the best approach is. As said, I think we'll be able to better judge with the implementation of http://jira.cubrid.org/browse/LETS-524 (and the addition of http://jira.cubrid.org/browse/LETS-586 logging and extensive testing).

For instance, being able to investigate all cases exposed in http://jira.cubrid.org/browse/LETS-511 will probably reveal some more scenarios which are not even covered by this implementation.